frr-exporter: init at 1.2.0

javaes opened the pull request:
## Description of changes
Prometheus exporter for FRR version 3.0+ that collects metrics from the FRR Unix sockets and exposes them via HTTP, ready for collecting by Prometheus.

Closes https://github.com/NixOS/nixpkgs/issues/309155

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

javaes committed:
prometheus-frr-exporter: init at 1.2.0

momeemt commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26506,6 +26506,7 @@ with pkgs;
   prometheus-fastly-exporter = callPackage ../servers/monitoring/prometheus/fastly-exporter.nix { };
   prometheus-flow-exporter = callPackage ../servers/monitoring/prometheus/flow-exporter.nix { };
   prometheus-fritzbox-exporter = callPackage ../servers/monitoring/prometheus/fritzbox-exporter.nix { };
+  prometheus-frr-exporter = callPackage ../by-name/fr/frr-exporter/package.nix { };
```
It can be built without adding prometheus-frr-exporter to pkgs/top-level/all-packages.nix because you have set the derivation file in pkgs/byname.

ref: https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md

javaes replied:
What about the apparent convention (couldn't find it documented) that all exporters for prometheus should start with `prometheus-[..]`? Should I rename the package and put it into the `by-name` path instead of the remapping here?

presto8 replied:
Moving the package into by-name/pr/prometheus-frr-exporter makes sense to me!

momeemt commented on pkgs/by-name/fr/frr-exporter/package.nix:
```diff
@@ -0,0 +1,33 @@
+  ];
+
+  meta = with lib; {
+    description = ''
+      Prometheus exporter for FRR version 3.0+ that collects metrics from the
+      FRR Unix sockets and exposes them via HTTP, ready for collecting by
+      Prometheus.
+    '';
```
```suggestion
    longDescription = ''
      Prometheus exporter for FRR version 3.0+ that collects metrics from the
      FRR Unix sockets and exposes them via HTTP, ready for collecting by
      Prometheus.
    '';
```

> `meta.description` must:
    Be short, just one sentence.
    Be capitalized.
    Not start with the package name.
    More generally, it should not refer to the package name.
    Not end with a period (or any punctuation for that matter).
    Provide factual information.
    Avoid subjective language.

ref: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

You should only put one sentence in `meta.description` and not end it with a period. You can use `meta.longDescription` instead. You must also set `meta.description` to follow these rules.

momeemt reviewed (changes requested):
Thank you for your contributions. I left two comments about `meta.description` and by-name. I would be grateful if you could confirm them.

isabelroses commented on pkgs/by-name/pr/prometheus-frr-exporter/package.nix:
```diff
@@ -0,0 +1,34 @@
+{ lib, buildGoModule, fetchFromGitHub }:
+
+buildGoModule rec {
+  pname = "prometheus-frr-exporter";
+  version = "1.2.0";
```
nit: https://nix.dev/guides/best-practices.html#recursive-attribute-set-rec
```suggestion
let
  version = "1.2.0";
in
buildGoModule {
  pname = "prometheus-frr-exporter";
  inherit version;
```

javaes commented:
@momeemt I addressed your comments. Thanks for the feedback!

@isabelroses thanks for your approval, is there anything else I can do to get this merged?

isabelroses commented:
> is there anything else I can do to get this merged?

Not that I'm aware of. But my best guess is to request a someone from nixpkgs committer team to review.

felbinger commented:
First of all, thanks for the package ;-)

We could, if you want to, also add the a prometheus module to this pr.
Something like this might work (not tested yet), I just copied it from [node exporter](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/monitoring/prometheus/exporters/node.nix) and adjusted some things.
```nix
# nixos/modules/services/monitoring/prometheus/exporters/frr.nix 
{ config, lib, pkgs, options, ... }:

let
  cfg = config.services.prometheus.exporters.frr;
  inherit (lib)
    mkOption
    types
    concatStringsSep
    concatMapStringsSep
    any
    optionals
    ;
  collectorIsEnabled = final: any (collector: (final == collector)) cfg.enabledCollectors;
  collectorIsDisabled = final: any (collector: (final == collector)) cfg.disabledCollectors;
in
{
  port = 9342;
  extraOpts = {
    enabledCollectors = mkOption {
      type = types.listOf types.str;
      default = [];
      example = [ "vrrp" ];
      description = ''
        Collectors to enable. The collectors listed here are enabled in addition to the default ones.
      '';
    };
    disabledCollectors = mkOption {
      type = types.listOf types.str;
      default = [];
      example = [ "bfd" ];
      description = ''
        Collectors to disable which are enabled by default.
      '';
    };
  };
  serviceOpts = {
    serviceConfig = {
      DynamicUser = false;
      RuntimeDirectory = "prometheus-frr-exporter";
      ExecStart = ''
        ${pkgs.prometheus-frr-exporter}/bin/frr_exporter \
          ${concatMapStringsSep " " (x: "--collector." + x) cfg.enabledCollectors} \
          ${concatMapStringsSep " " (x: "--no-collector." + x) cfg.disabledCollectors} \
          --web.listen-address ${cfg.listenAddress}:${toString cfg.port} ${concatStringsSep " " cfg.extraFlags}
      '';
    };
  };
}
```
For now I removed the "systemd hardening" (RestrictAddressFamilies, ...), but we should think about it.

johannwagner commented:
@WOBCOM might be also interested in this, probably a merge from @yu-re-ka possible? 👉👈
